gapic/src/main/com/google/gapid/views/DeviceValidationView.java

```diff
@@ -15,175 +15,304 @@
+    // validation result.
+    retryButton.addListener(
+        SWT.Selection,
+        e -> {
+          ValidateDevice(device);
+        });
+
+    models.devices.validateDevice(device, this::displayResult);
+  }
+
+  private void ClearResults() {
+    statusLoader.setVisible(false);
+    statusText.setVisible(false);
     retryButton.setVisible(false);
+    removeAllSelectionListeners(retryButton);
+    removeAllSelectionListeners(statusText);
+
+    traceValidationFailureCount = 0;
+    passedValidation = false;
+
     if (errorMessageGroup != null) {
       errorMessageGroup.dispose();
       errorMessageGroup = null;
       helpComposite.dispose();
       helpComposite = null;
       requestLayout();
     }
+  }
 
-    DeviceValidationResult cachedResult = models.devices.getCachedValidationStatus(dev);
-    if (cachedResult != null) {
-      if (cachedResult.passed || cachedResult.skipped) {
-        setValidationStatus(cachedResult);
-        return;
-      }
+  private void displayResult(DeviceValidationResult result) {
+    boolean passedOrSkipped = result.passedOrSkipped();
+    statusLoader.stopLoading();
+    statusLoader.updateStatus(passedOrSkipped);
+    passedValidation = passedOrSkipped;
+
+    statusText.setText(convertToMessage(result.errorCode));
+    setStatusTextLink(result.errorCode);
+
+    if (passedOrSkipped) {
+      notifyListeners(SWT.Modify, new Event());
+      return;
     }
 
-    statusLoader.startLoading();
-    statusText.setText("Device support is being validated");
+    helpComposite =
+        createComposite(this, withMarginAndSpacing(new GridLayout(1, false), 0, 0, 0, 0));
+    helpComposite.setLayoutData(withSpans(new GridData(), 3, 1));
+    addExplanationMessage(helpComposite, result.errorCode);
+
+    switch (result.errorCode) {
+      case FailedPrecondition:
+        // Intentionally empty.
+        break;
+      case FailedTraceValidation:
+        // Only show the retry button for non-deterministic errors.
+        retryButton.setVisible(true);
 
-    // Assign appropriate callback for retry button.
-    for (Listener listener : retryButton.getListeners(SWT.Selection)) {
-      retryButton.removeListener(SWT.Selection, listener);
+        traceValidationFailureCount++;
+        if (traceValidationFailureCount >= TRACE_VALIDATION_FAILURE_COUNT) {
+          addFileBugSection(helpComposite, result.tracePath);
+        }
+        break;
+      default:
+        addFileBugSection(helpComposite, result.tracePath);
+        break;
     }
-    retryButton.addListener(SWT.Selection, e -> {
-      ValidateDevice(dev);
-    });
 
-    models.devices.validateDevice(dev, this::setValidationStatus);
-  }
+    // Separate error text from help text as a workaround to enable proper text wrapping.
+    errorMessageGroup =
+        withLayoutData(
+            createGroup(this, ""), withSpans(new GridData(SWT.FILL, SWT.TOP, true, false), 3, 1));
+    Text errText =
+        withLayoutData(
```

Comment:
Nit: Is this line break after assignment a java style?

Reply:
According to the Google Java style guide:
`When a line is broken at an assignment operator the break typically comes after the symbol, but either way is acceptable.`

I think there's nothing prohibiting `withLayoutData` to be on the same line as `Text errText =` but I ran the java formatter and this is what it produced